### ISSUE TITLE
If segmented by goal other goal conversations don't show up

### ISSUE THREAD

**Issue by Akuukis** [piwik/piwik#8854]
**Problem**: When I segment by `"Visit converted a specific goal ID";"is";2` then all other goals show up as converted 0 times, despites that in visitor log I can see that visitors convert multiple goals.

**Context**: I have a lot of goals and people usually convert several per visit. So I want to see what other goals they convert if they have converted goal ID 2.

**Comment by sidazhou** [comment]
I'm troubled with the same issue. Have you gotten any further yet?

**Comment by Akuukis** [comment]
No :(

**Comment by mattab** [comment]
Thanks for the report!

**Comment by tomasztomik** [comment]
@mattab I encountered exactly the same issue recently. 

It seems to be caused by the fact that segment is filtering 'per-goal' not 'per-visit'. Meaning if the condition is set to goalId=7, it will show only conversion where goalId was 7 (not all goals from the visit, where at least one goal id was = 7).

Such behavior seems to be a result of limitation of how segments are created. In reports for each goal there is a SQL query that has part `idgoal=<id of a goal>`. While browsing segment, another idgoal part is added, so we get for example `idgoal=20 AND idgoal=7`, which never will be true.

Do you have an idea how can it be fixed?

Especially that use case described in the issue's description (I have a lot of goals and people usually convert several per visit. So I want to see what other goals they convert if they have converted goal ID 2.) is very common, so would be nice to have it fixed soon.

Regards,
Tomasz

**Comment by tsteur** [comment]
I'm not sure if I created an issue for it a while ago or not. It is definitely a general bug that also applies to action segments etc and explains why often "no data" appears in reports when segmenting depending on whether an action, visit or goal segment is applied to a specific report. We could fix it by adjusting the segmentation queries but it's a bit of work, especially to make sure it still works afterwards

**Comment by tomasztomik** [comment]
@tsteur @mattab Do you have any idea if it can be fixed in the near future, or is it rather long term topic for you at this moment? I am aware that it requires a lot of work, however, it's a bug in one of the critical functionalities on the other hand.

**Comment by tsteur** [comment]
Yes It is critical functionality and it would be good to fix the segmentation. It's right now not a priority unless someone wants to sponsor it because it is 1) complex 2) we need big database to test it as it may 3) lower performance of segments because it may result in more joins or subqueries. Ideally we would want to find a way to not regress performance for segments.

**Comment by mattab** [comment]
Hi @tomasztomik as @tsteur explained it is a must-fix bug we can and would like to solve, it would likely take one day or two to fix so haven't take the time yet. Increasing priority of the issue

**Comment by tsteur** [comment]
@mattab I'm not sure if you are talking about only fixing this issue or the general segmenting issue. It already took me 2 days to refactor just some parts of the segmenting logic and changing the segmenting to make them work proerply might be more like 3-5 days with ensuring everything still works, keeping the queries fast somehow, writing plenty of new tests etc.

**Comment by mattab** [comment]
@tsteur was referring to fixing this specific issue with the segment `Visit converted a specific goal ID` in one or maybe two days, would be good first step
